client/consensus/pow/src/worker.rs

```diff
@@ -0,0 +1,213 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{pin::Pin, time::Duration, collections::HashMap, any::Any, borrow::Cow};
+use sc_client_api::ImportNotifications;
+use sp_runtime::{DigestItem, traits::Block as BlockT, generic::BlockId};
+use sp_consensus::{Proposal, BlockOrigin, BlockImportParams, import_queue::BoxBlockImport};
+use futures::{prelude::*, task::{Context, Poll}};
+use futures_timer::Delay;
+use log::*;
+
+use crate::{INTERMEDIATE_KEY, POW_ENGINE_ID, Seal, PowAlgorithm, PowIntermediate};
+
+/// Mining metadata. This is the information needed to start an actual mining loop.
+#[derive(Clone, Eq, PartialEq)]
+pub struct MiningMetadata<H, D> {
+	/// Currently known best hash which the pre-hash is built on.
+	pub best_hash: H,
+	/// Mining pre-hash.
+	pub pre_hash: H,
+	/// Pre-runtime digest item.
+	pub pre_runtime: Option<Vec<u8>>,
+	/// Mining target difficulty.
+	pub difficulty: D,
+}
+
+/// A build of mining, containing the metadata and the block proposal.
```

Comment:
I don't understand what "build of mining" means; can you elaborate on what it means? `MiningProposal` perhaps, but it's sort of taken!

Reply:
It's basically a combination of metadata and the block proposal. Metadata is provided to miner for generating the seal, it then combined back with the block proposal to finally "build" the block. So I called it "build", but maybe there should be a better name.

Reply:
I want to merge this PR in the next few hours. Don't think this naming issue is blocking. If we cannot think of a better name now, we can change it in the future!